### DISCUSSION TITLE
Time derivative of Time-varying parameters possible?

### DISCUSSION THREAD

**Discussion by Doctor-N0** [do-mpc/do-mpc#271] [General]
I am using Time-varying parameters for tracking a reference trajectory in a very similar approach as the [documentation](https://www.do-mpc.com/en/latest/FAQ.html).

To be specific, I use Set Points in x and y direction:
```
# Introduce tvp: Set-point
x_set_point= model.set_variable(var_type='_tvp', var_name='x_set_point')
y_set_point= model.set_variable(var_type='_tvp', var_name='y_set_point')
```
for my objective function
```
lterm = (model.x['x']-model.tvp['x_set_point'])**2 + (model.x['y']-model.tvp['y_set_point'])**2
mpc.set_objective(mterm=mterm, lterm=lterm)
```
The setpoints can be a straight line, a Lissajous Curve or some other shape:
```
def tvp_fun_mpc(t_now):
    # straight line
    # tvp_struct_mpc['_tvp',:,'x_set_point'] =  0.5*t_now
    # tvp_struct_mpc['_tvp',:,'y_set_point'] =      t_now
    
    # Lissajous Curve
    tvp_struct_mpc['_tvp',:,'x_set_point'] = 2 * sin(0.5*t_now)
    tvp_struct_mpc['_tvp',:,'y_set_point'] = 2 * sin(1*t_now)
    return tvp_struct_mpc
```

Which works quite well. 

Now I would like to calculate the reference angle "theta_set":
![grafik](https://user-images.githubusercontent.com/105864309/170980572-1e1c6b19-ff9d-4a0e-8fbd-b068a90ce381.png)
to use it in the lterm:
```
lterm = (model.x['x']-model.tvp['x_set_point'])**2 + (model.x['y']-model.tvp['y_set_point'])**2 - (model.x['theta']-model.tvp['theta_ref'])**2
```
However, I don't know if it is even possible to calculate the time derivative of a Time Varying Parameter in Do-MPC? 
Thank you for your help!

**Comment by 4flixt** [comment]
Can you maybe introduce the time derivative of the time-varying parameter as a time-varying parameter? And then introduce your setpoints as states with the newly introduced tvp as their dynamics?

**Reply by Doctor-N0** [reply]
If I understood you correctly, you're suggesting to introduce new time-varying parameters "`dx_r`", "`dy_r`" and new states `x_r` and `y_r`. Then "implement" the time derivative via `model.set_rhs('x_r',dx_r ).` But how do I tell do-mpc that it should do the derivation `dx_r = d/dt (x_r)`? Maybe using the `numpy.diff()` command?

**Reply by 4flixt** [reply]
You understood me correct. Why would you need to do the derivation now? You have introduced ``dx_r`` and ``dy_r`` so you have access to the time derivative of the ``tvp``.
